packages/itwin/tree-widget/src/components/trees/models-tree/internal/ModelsTreeVisibilityHandler.ts

```diff
@@ -241,9 +276,13 @@ class ModelsTreeVisibilityHandlerImpl implements HierarchyVisibilityHandler {
         if (elements?.size) {
           observables.push(
             from(elements).pipe(
-              mergeMap((key) => {
-                const { modelId, categoryId, elementId } = parseElementKey(key);
-                return this.getElementDisplayStatus({ modelId, categoryId, elementId });
+              releaseMainThreadOnItemsCount(50),
+              mergeMap(([categoryKey, elementIds]) => {
+                const { modelId, categoryId } = parseCategoryKey(categoryKey);
+                return from(elementIds).pipe(
+                  releaseMainThreadOnItemsCount(1000),
+                  mergeMap((elementId) => this.getElementDisplayStatus({ modelId, categoryId, elementId, ignoreTooltip: true })),
+                );
               }),
             ),
```

Comment:
Do we need both `releaseMainThreadOnItemsCount` calls here? It seems the first one ensures we don't attempt to call `parseCategoryKey` for every element node in a single task. Then we call `getElementDisplayStatus` in batches of `1000` elements. Did you try changing the top-level batch size to `1000` and removing the inner `releaseMainThreadOnItemsCount`?

Also, maybe it makes sense to add a limit to how many tasks in parallel are launched my `mergeMap`?

Reply:
Tried using `mergeMap` second parameter instead of `releaseMainThreadOnItemsCount(50)` before but I saw the we get at least 1 longer running microtask >50ms (I guess it's for the root subject visibility as it has 87k filtered elements underneath)